### PR TITLE
[POC][docs] Add iteration on the anatomy demo

### PR DESCRIPTION
**Pull request by danilo-leal**
- [x] I have followed (at least) the [PR section of the contributing guide](https://github.com/mui/base-ui/blob/HEAD/CONTRIBUTING.md#sending-a-pull-request).

Riffing on https://github.com/mui/base-ui/pull/267 — not meant for merge.

https://github.com/mui/base-ui/assets/67129314/b7ef0f03-54d8-4f37-bfdc-6a0cc472187f

### CODE REVIEW

**Comment by atomiks** [comment]
I really like these, but I think the appearance should be "wireframe"-y, since in its current state, it kind of looks like you should be able to use the component (but it's `readonly`).

Like this, but with the style you've made:

Screenshot 2024-04-06 at 8 59 48 AM

**Comment by danilo-leal** [comment]
That's a great idea — cc @zanivan. I was wondering about ways to solve the fact that the "parts highlighters" (let's call it like this 😅) make the component non-interactive. The wireframe-y aspect could already remove that expectation 👍

**Comment by colmtuite** [comment]
There are a few issues with this UI:

1. The elements have interactive styles, but they're not interactive.
2. It's not usable with a keyboard: I can't tab through the elements or use arrow keys.
3. It's not usable on touch devices, since the elements are not interactive.

What is the core problem that this work is trying to address? Is this part of a larger docs update effort?

We could do something like RA with the wireframe images, but they're not accessible to blind people either people you can't really communicate the image in the `alt` attr. I think it would be difficult to improve this section of the docs as it exists currently. Perhaps we could add a simple `// optional` code comment beside optional parts.

**Comment by danilo-leal** [comment]
> What is the core problem that this work is trying to address? Is this part of a larger docs update effort?

We're exploring a more interactive/visual way for readers to understand a given component's anatomy—which part corresponds to what, visually. And nah, we're just prototyping here. I've personally had this idea for a long time.

To the points you listed:

1. You mean the Number Field, right? If yes, that's what we were talking about up there: making it appear less like a "full" demo (not reusing the intro demo styles) and more wireframe-y to convey better that it's not supposed to be interactive.
2. We can definitely improve that (+ the code) as needed—we're just exploring here.
3. Same solution as point one, I guess? The actual component in this demo isn't meant to be interactive.

**Comment by zanivan** [comment]
I agree with Danilo on this one. This is meant to be a visual aid for the anatomy grouping/hierarchy. There's definitely room for improvement, but I think it'd be a good thing to offer in the docs.

I like the idea of making the components look more like wireframes in these demos! I tried to make it blue, following RA example, and gray, like common wireframes—also used IBM Plex Mono for it, with more 'specific' labels. What do you think?

image

image

**Comment by zanivan** [comment]
Iterated on it on #267, on [f45e717](https://github.com/mui/base-ui/pull/267/commits/f45e7174c95391fccd5e1a3cf3ffdd9227c2c69a):


https://github.com/mui/base-ui/assets/37222944/e6a602aa-38de-4f72-a195-924b74d068be

- added focus style to the labels, so it's possible to navigate the demos with tab.
- styled demo and its container to resemble a wireframe.
- brought the code snippet with the anatomy back.

I think these mitigate almost all the cons we had.

**Comment by atomiks** [comment]
I'd go full wireframe (only borders, no background colors). It should look as non-interactive and picture-like as possible. I'd then remove the `disabled` cursor style when hovering it, having it only be `cursor: default` (like an `` tag).

Screenshot 2024-04-09 at 9 44 09 AM

Then the "hover" borders would possibly be red?

**Comment by colmtuite** [comment]
@danilo-leal @zanivan 

**a11y**
> To the points you listed:
> You mean the Number Field, right?

For these 3 points, I was talking about the code bubble thingys. They render `<span>`, so they have interactive styles but they're not interactive, which is confusing for all users and unusable for touch devices.

Since then, the code bubble thingys can now receive keyboard focus. But how would I use this feature on a touch device? It's very important that any UI we add to Base UI docs is accessible.

**What is the problem we're solving?**
> We're exploring a more interactive/visual way for readers to understand a given component's anatomy—which part corresponds to what, visually

Is there a known problem here though? It seems pretty clear to me which part does what. Like it's obvious which part is the increment and decrement buttons, and it's obvious what those buttons do. It's less obvious what parts like `Group` do, but i don't think this new feature actually helps with that? Sure, I can see the `Group` part being highlighted, but that doesn't help me understand what it does? So I don't understand what value this features provides that a simple anatomy code block doesn't provide on its own.

**Suggestion to move this UI to the code block**
If we were to go with this kind of interactive bubble hover/click UI, I'd consider having it be a code block feature instead, perhaps where you hover a code line and it highlights that part of the UI in a live preview. A couple reasons:

1. The current example duplicates the code structure from the code block, so I'm looking at the same code twice, once in the code block, and again in the side-by-side interactive anatomy UI. It seems confusing to me, and also consumes precious docs space unnecessarily imo.
2. The code block is already inaccessible to some degree (which is unavoidable), so moving this feature to the code block is less problematic.

**Problem with hidden content**
The NumberField renders all of its parts on page load. But how would this UI work for components that have unmounted parts, like Popover, Menu, Select, Dialog etc.? The Popover is closed by default, so when I hover the `Popup` part, how would I see the highlighted area?

**Comment by danilo-leal** [comment]
> but they're not interactive,

I don't understand this point here—what do you mean they're _not_ interactive? Hovering/focusing/pressing them and then having another thing appear/be highlighted doesn't count as interactivity?

> What is the problem we're solving?

That's a valid question and valid points. I agree it doesn't immediately add new information compared to a simple codeblock. I think the biggest value is that it is just more visual than code, which might be nice for more visual folks. It may be worth it to get the community's input about whether this would make sense and what more to refine here. React Aria has anatomy images with a slightly different approach than what we're experimenting with here (they're not using the exact component part code name), and I've always found it quicker to understand a given component's anatomy that way.

Screenshot 2024-04-09 at 08 52 26



> Suggestion to move this UI to the code block

Yeah, we thought about this, too! I was bothered by the fact that, in my demo, I've indented the boxes to resemble code, but I was not closing the tags 😬 It's just a bit more complex to pull off in prototype mode, but it would potentially be better, indeed.

> Problem with hidden content

I'd think that for those other components, we'd just open them up on the anatomy demo for the sake of demonstration?

**Comment by colmtuite** [comment]
> what do you mean they're not interactive?

I mean `<span>` is not an interactive element, so it introduces a11y concerns. For example, touch devices don't support hover or tabbing, and I can't click this element because it's not an interactive element, so the feature becomes unusable on most devices.

> React Aria has anatomy images

Yep I've seen RA's approach. Seeing that actually left me wondering the same thing when I first saw it: I already know what most of the parts are/do just by reading the component name, like `Trigger`, `Popover`, `MenuItem`, `MenuItemKbd`, `MenuItemLabel`, `MenuItemDescription`. These parts all intuitive just by reading the component name imo.

There are a few parts that might leave me wondering what they are/do, like `Menu`, `Section`, and `SectionHeader`. But highlighting them with this UI wouldn't always give me more information. Sometimes it would, like with `SectionHeader` for example, pointing to the text label above a section would indicate what this component is for. But often times, like with `Menu` and `Section`, I'd still be left in the dark, because I just see a large area being highlighted, I still I don't what the thing does, or for example what the difference is between `Menu` and `Popover`.

To have my questions answered, I think I would actually need some sort of info icon button UI that opens a `Popover` on click. So for example, I click the info icon button beside `Menu` and inside its Popover, I can read "`<Menu>` renders a `<div>` that gets mounted/unmounted based on the component's open/closed state. The `<Popover>` component goes inside this."

So to summarise, I have two thoughts:

1. I'm not convinced that there is a significant problem to be solved here. I think the purpose of each component is intuitive for the most part, just by reading the component name.
2. For many of the cases where the purpose of a component is not entirely intuitive just be reading its name, I don't think a diagram (whether interactive or not) would help.

It seems to me that the best options here are:

1. Do nothing, just have a code block to show the structure of the component.
2. Add code comments beside each component in the code block to give a bit more insight into its purpose.

**Comment by danilo-leal** [comment]
> I mean <span> is not an interactive element,

Gotcha, appreciate the clarity. The debate around the value for the reader/problem we're solving is more important to me for now, though. Depending on where that goes, we can fine-tune the implementation so it meets all of the accessibility requirements, making sure we don't damage Base's reputation in any way.

>  I can read "`<Menu>` renders a `<div>` that gets mounted/unmounted based on the component's open/closed state. The <Popover> component goes inside this."

I think we can prototype that! Instead of just showing the highlight on top of the component "wireframe," we could also display a little explanation description to see how it feels like.

> It seems to me that the best options here are:

We're just playing around for the time being. I think it's cool to explore other routes, try to expand this idea, etc. Just this conversation alone has generated some new thoughts! Then, we can evaluate if it makes sense to generalize this, refine the code, and then apply it to the actual documentation. No strings attached at the moment. It's okay if we just throw it away and don't ever use it.

> I don't think a diagram (whether interactive or not) would help.

That's the part I'm most curious about. For simpler components, for sure (e.g., React Aria doesn't have an illustrated image on the Button component). But for more complex ones, maybe a more visual way to consume this information can be more helpful.
